### ISSUE TITLE
Model Serialization

### ISSUE THREAD

**Issue by justin1121** [OpenMined/OpenMined#296]
We need to be able to serialize models in OpenMined. This needs to be done to send models over the wire and potentially to IPFS. Probably a few different ways to achieve this. 

Native C#: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/serialization/

Pros: Built in
Cons: Is this supportable across languages?

Protobufs: https://github.com/google/protobuf

Pros: Portable
Cons: Third party dependency

ONNX: https://github.com/onnx/onnx (this one is probably not an option for now, just putting it here for discussion sake)

Pros: Will be handy to have this one day so we can import and export models to other frameworks
Cons: A lot of heavy lifting left to do to support this

I don't know much about any of these options so feel free to add/correct anything! 😄

**Comment by bendecoste** [comment]
imo protobuffs would be worth looking into especially if there is also going to be a .js framework

**Comment by justin1121** [comment]
I think to start we're just doing unity -> unity deserialization/serialization. So builtin C# might be a good idea. But if there's a slight chance we want to support other languages probably best to start with something portable so we don't have to make it portable in the future.

**Comment by bendecoste** [comment]
we already have to do serialization for to_numpy and it comes up a bit during development (agree it's less needed when you are actually training)

It would also be beneficial for the `Command` object 
```c#
public class Command
	{
		// given that SyftController keeps lists of objects of base types
		// (at the time of writing this is only Tensors) then this command
		// selects one of these generic types and performs a command.
		public string objectType;
		// i.e. "tensor"
		public int objectIndex;
		//

		// name of the function to be called
		public string functionCall;

		public float[] data;
		public int[] shape;

		public string[] tensorIndexParams;
	}
```

**Comment by morgangiraud** [comment]
Maybe this format would be interesting: [Flatbuffers](https://google.github.io/flatbuffers/)
- [Github](https://github.com/google/flatbuffers)

It is a light Protobuff, simpler and performance oriented. It works on Python/JS/C# already and many more to comes.

**Comment by morgangiraud** [comment]
# Some thoughts on serialisation

## Graph serialisation
ML frameworks deals with graphs containing operations and weights(data). To serialise it, one has to:
- Save the model architecture
- Save the weights of the model
- Save the potential optimiser/loss/etc. architecture
- Save the optimiser/loss/etc. states

## Choice made by other frameworks

### Keras
#### single file
- Format:  HDF5 
- Usage:
  - model.save('my_model.h5')
  - load_model('my_model.h5')
- File structure:
```
{
    keras_version:str
    backend:str
    model_config:str (JSON)
    model_weights: HDF5.group {
        weight_names: [str]
        ...weight_name...:weight_values
    }  
    layer_names:str
      ...layers...:HDF5.group
        weights_names:str
        ...weight_name:value...:HDF5.dataset
    training_config: object {
        Optimizer/loss/metrics/etc. 
    }
}
```
#### model architecture
- Format: JSON/YAML
- Usage:
  - model.to_json()
  - model = model_from_json()
- Example
    ```{"class_name":"Sequential","config":[{"class_name":"Dense","config":{"name":"dense_5","trainable":true,"batch_input_shape":[null,784],"dtype":"float32","units":512,"activation":"linear","use_bias":true,"kernel_initializer":{"class_name":"VarianceScaling","config":{"scale":1.0,"mode":"fan_avg","distribution":"uniform","seed":null}},"bias_initializer":{"class_name":"Zeros","config":{}},"kernel_regularizer":null,"bias_regularizer":null,"activity_regularizer":null,"kernel_constraint":null,"bias_constraint":null}}],"keras_version":"2.1.2","backend":"tensorflow"}```

#### Model weights
- Format : HDf5
- Usage:
  - model.save_weights('my_model_weights.h5')
  - model.load_weights('my_model_weights.h5')

### PyTorch

#### single file
- Format:  pickle 
- Usage:
  - torch.save(the_model, PATH)
  - the_model = torch.load(PATH)
- File structure:
```
{
    MAGIC_NUMBER
    PROTOCOL_VERSION
    sys_info
    model with persistent_id function
    serialized_storage_keys
}
````

#### model architecture
I didn't see any way to dump the model architecture by itself, but I'm a noob on PyTorch so feel free to correct me.

#### Model weights
- Format: Pickle
- Usage:
  - torch.save(the_model.state_dict(), PATH)
  - the_model = TheModelClass(*args, **kwargs)
  - the_model.load_state_dict(torch.load(PATH))
  

### Tensorflow
Format for everything: Protobuf

#### Single file
You can do it in Tensorflow, but it has the disadvantage of freezing the variables, this is an irreversible process.
- Format: Protobuf

#### Model architecture and Model weights
- Format: Protobuf
- Usage:
The usage is a little bit more complex since you need to launch a Session to load the weights
  - design the graph
  - Start a session
  - Save the graph and the weights

## Which format for OpenMined
OpenMined has some very specific characteristics (IPFS storage, highly distributed) which put pressure on on the serialization size (network and IPFS load). The graph serialisation should at least, satisfy these points:
  - the weights serialization should be memory efficient (less network/blockchain pressure)
  - make the to_numpy func fast and memory efficient (fast numpy conversion directly from the binary format could be great)
  - Works potentially everywhere (C#/JS/Python)

Given the previous framework serialisation description, we have:
  - PyTorch and Pickle:  python only
  - Keras and the JSON/HDF5 duo: accessible but HDF5 seems not very efficient and 
  - Protobuf, satisfy almost all constraint: not really memory efficient.
  - Flatbuffer: basically. a memory efficient Protobuf 
  - ONNX: I have no idea yet on this one

### Potential next step
  - Serialise everything as Flatbuffer (serialize the graph as a JSON in Flatbuffer, and the weights directly), it's a start and if we need, we can extend it to the Protobuff format easilly
  - Make save/load (with weights) work inside OpenMined only
  - See how all this behaves/integrates and Discuss where to go from this point.

### Other questions of interest
**This part is a little bit messy, it's more like random remarks that I wanted to share. Feel free to tell me to edit (or remove) those remarks**
- interopability with PyTorch and Keras (and so TensorFlow): we will need a Hdf5 and a pickle transformer in PySyft. Special care should be done when choosing names of layers, functions, etc  in this case.
- Not sure how we want to handle graph topology. So far, the line between a node, variables, layer or a model is quite blurry. Is there any work on this?
- Advantage of serialization: The summary function is heavy on the network (3*nb_layer + 1 calls) right now (just like the current Keras load_model implementation) , when the topology serialization will be over, we could retrieve the summary in only one network-call function
- While implementing this hack, I was wondering if it would be easier to fork Keras and build a OpenMined backend in it than rebuild an entire interface to mimick the Keras one from the PySyft repo?

The 2 WIP PR:
[PySyft](https://github.com/OpenMined/PySyft/pull/753)
[OpenMined](https://github.com/OpenMined/OpenMined/pull/314)

What do you think @iamtrask @justin1121 @bendecoste?

**Comment by justin1121** [comment]
This is great! Good idea figuring out how pytorch, etc. do it. Not sure the benefit of using flatbuffer over just JSON is. @bendecoste has spent some time inserting a json blob into IPFS and that seems to work.

> interopability with PyTorch and Keras (and so TensorFlow): we will need a Hdf5 and a pickle transformer in PySyft. Special care should be done when choosing names of layers, functions, etc in this case.

This part I'm not sure we really need to support. This is the point of ONNX its a cross-platform protocol that hopefully all deep learning frameworks would support one day. So if we wanted to load models in and out of PySyft to keras/PyTorch then we would use that. It would be a lot of effort to support decoding/encoding to all the platforms.

**Comment by iamtrask** [comment]
So on the one hand, ONNX is a great format with lots of veristility. However, on the other hand, we're supporting entire interfaces for PyTorch and Keras, and the ability to go back and forth in the same notebook is pretty freaking sweet. Also, going direct is likely to be quite a bit faster. It might even be useful for unit testing.

Also, the author of Keras makes some good points here. ONNX doesn't have many users yet - but we _know_ that people will want to go back and forth between OpenMined and Keras/Pytorch. Furthermore, we could use this as a gateway to ONNX if they support it.

That being said - PyTorch already has support for ONNX (http://pytorch.org/docs/master/onnx.html) so we could also make the argument that we should just use ONNX in the backend to convert between OpenMined and PyTorch.

Lots of great options! Thoughts?

**Comment by iamtrask** [comment]
I'd say the highest priority, however, is very _fast and efficient_ serialization so that we can send models over the wire. Any guesses which will be best for that?

**Comment by bendecoste** [comment]
i think serialization efficiency will be pretty negligible compared to waiting for someone to pick up your job & actually running the training.  Not to say that we should ignore the performance

**Comment by morgangiraud** [comment]
My research on this shows that the subset of Protobuf: flatbuffer seems to be the fastest and most efficient.

But we might try something like keras where we serialize the topology in JSON and in Flatbuffer/protobuf the weights. The JSON could then be furthermore encoded in the flatbuffer/protobuff as a simple string.
Not sure if that is the best way, but probably the fastest one to implement.
